src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -31,11 +31,113 @@ public struct CallGraphNode
             public QsNullable<ImmutableArray<ResolvedType>> TypeArgs;
         }
 
+        // TODO: 
+        // This is the method that should be invoked to verify cycles of interest,
+        // i.e. where each callable in the cycle is type parametrized.
+        // It should probably generate diagnostics; I'll add the doc comment once its use is fully defined. 
+        internal static bool VerifyCycle(CallGraphNode rootNode, params CallGraphEdge[] edges)
+        {
+            var parent = rootNode.CallableName;
+            bool EqualsParent(QsQualifiedName origin) => origin.Namespace.Value == parent.Namespace.Value && origin.Name.Value == parent.Name.Value;
+            var validResolution = TryCombineTypeResolutions(parent, out var combined, edges.Select(edge => edge.ParamResolutions).ToArray());
+            var resolvedToConcrete = combined.Values.All(res => !(res.Resolution is ResolvedTypeKind.TypeParameter tp) || EqualsParent(tp.Item.Origin));
+            return validResolution && resolvedToConcrete;
+            //var isClosedCycle = validCycle && combined.Values.Any(res => res.Resolution is ResolvedTypeKind.TypeParameter tp && EqualsParent(tp.Item.Origin));
+            // TODO: check that monomorphization correctly processes closed cycles - meaning add a test...
+        }
+
+        /// <summary>
+        /// Given a sequence of type resolutions specifying e.g. subsequent concretizations as part of a nested expression, 
```

Comment:
I have no idea; what seems more like a valid and accurate English word? I was just assuming the verb concretize exists and the corresponding noun would be concretization.

Reply:
In my experience, putting a -tion at the end of a verb makes a noun that describes the process of performing that verb, not necessarily the outcome of performing the verb, although, because English is pretty inconsistent, I am certain there are plenty of counter examples.

Concretization seems like it is referring to the process of performing the verb 'concretize' rather than the outcome.

Reply:
Yes, but that refers to the verb *to concrete*, which also seems to be primarily used for physical processes. The abstract use seems uncommon. For example, the [concretion](https://en.wikipedia.org/wiki/Concretion) Wikipedia article is only about physical concretions, while [concretization](https://en.wikipedia.org/w/index.php?title=Concretization&redirect=no) redirects to an article about the concretization of abstract ideas.